src/cli/util.js

```diff
@@ -397,33 +399,129 @@ function createIgnorerFromContextOrDie(context) {
   }
 }
 
-function eachFilename(context, patterns, callback) {
-  // The '!./' globs are due to https://github.com/prettier/prettier/issues/2110
-  const ignoreNodeModules = context.argv["with-node-modules"] !== true;
-  if (ignoreNodeModules) {
-    patterns = patterns.concat(["!**/node_modules/**", "!./node_modules/**"]);
+/**
+ * Get stats of a given path.
+ * @param {string} filePath The path to target file.
+ * @returns {fs.Stats|undefined} The stats.
+ * @private
+ */
+function statSafeSync(filePath) {
+  try {
+    return fs.statSync(filePath);
+  } catch (error) {
+    /* istanbul ignore next */
+    if (error.code !== "ENOENT") {
+      throw error;
+    }
   }
-  patterns = patterns.concat(["!**/.{git,svn,hg}/**", "!./.{git,svn,hg}/**"]);
+}
 
-  try {
-    const filePaths = globby
-      .sync(patterns, { dot: true, nodir: true })
-      .map(filePath => path.relative(process.cwd(), filePath));
+/**
+ * Check if a string is a glob pattern or not.
+ * @param {string} pattern A glob pattern.
+ * @returns {boolean} `true` if the string is a glob pattern.
+ */
+function isGlobPattern(pattern) {
+  return isGlob(path.sep === "\\" ? pattern.replace(/\\/gu, "/") : pattern);
+}
+
+const globbyOptions = { dot: true, nodir: true, absolute: true };
+function eachFilename(context, maybePatterns, callback) {
+  const withNodeModules = context.argv["with-node-modules"] === true;
+  // TODO: use `ignore` option for `globby`
+  const extraPatterns = [
+    // The '!./' globs are due to https://github.com/prettier/prettier/issues/2110
+    ...(withNodeModules ? [] : ["!**/node_modules/**", "!./node_modules/**"]),
+    "!**/.{git,svn,hg}/**",
+    "!./.{git,svn,hg}/**"
+  ];
+
+  const allFilesInPatterns = ["**/*", ...extraPatterns];
+  // Ignores files in version control systems directories and `node_modules`
+  const ignoredDirectories = [".git", ".svn", ".hg"];
+  if (!withNodeModules) {
+    ignoredDirectories.push("node_modules");
+  }
+
+  let files = [];
+  const cwd = process.cwd();
+  const patterns = [];
+
+  for (const pattern of maybePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+    const stat = statSafeSync(absolutePath);
+    if (
+      stat &&
+      path
+        .relative(process.cwd(), absolutePath)
+        .split(/[\\/]/)
+        .some(directory => ignoredDirectories.includes(directory))
+    ) {
+      continue;
+    }
 
-    if (filePaths.length === 0) {
+    if (
+      stat &&
+      stat.isDirectory() &&
+      // `dot pattern` and `expand directories` support need handle differently
+      // for backward compatibility reason only expand `directories` like a glob pattern
+      // see https://github.com/prettier/prettier/pull/6639#issuecomment-548949954
+      isGlobPattern(pattern)
+    ) {
+      files = [
+        ...files,
```

Comment:
What's wrong with `files.push(...`? Why reassign the variable?

Reply:
I considered use that, but is there arguments length limit in this? I'm  not sure.

Reply:
That's an interesting concern. Indeed, the spread syntax seems to be able to handle more elements. On my machine:
```console
> [].push(...Array(1e5))
100000
> [].push(...Array(2e5))
RangeError: Maximum call stack size exceeded
> [...Array(2e5)]
[ undefined,
  undefined,
  undefined,
...
```
Okay, let's keep it this way.

Reply:
This is also why people don't use `String.fromCharCode.apply(String, codePoints)` to build string.

Alternate way, loop and `.push(...x.slice(1e5))` one time.